File: frontend/src/types/metadata.ts
/* ─────────────────────────────────────────────
   Metadata Types (from backend /api/metadata)
───────────────────────────────────────────── */

export type SemanticType =
  | 'categorical'
  | 'numeric'
  | 'temporal'
  | 'identifier'
  | 'measure'
  | 'unknown';

export type RelationshipKind = 'many_to_one' | 'one_to_one' | 'many_to_many';

export type MetadataJobStatus = 'pending' | 'running' | 'succeeded' | 'failed' | 'cancelled';

export interface TableMetadata {
  user_id: string;
  table_name: string;
  description: string | null;
  primary_date_column: string | null;
  grain: string | null;
  generated_at: string | null;
  edited_by_user: boolean;
  created_at: string | null;
  updated_at: string | null;
}

export interface TableMetadataPatch {
  description?: string;
  primary_date_column?: string;
  grain?: string;
}

export interface ColumnMetadata {
  user_id: string;
  table_name: string;
  column_name: string;
  data_type: string;
  semantic_type: SemanticType;
  description: string | null;
  is_filterable: boolean;
  cardinality: number | null;
  sample_values: unknown[] | null;
  generated_at: string | null;
  edited_by_user: boolean;
  created_at: string | null;
  updated_at: string | null;
}

export interface ColumnMetadataPatch {
  semantic_type?: SemanticType;
  description?: string;
  is_filterable?: boolean;
}

export interface Relationship {
  user_id: string;
  from_table: string;
  from_column: string;
  to_table: string;
  to_column: string;
  kind: RelationshipKind;
  confidence: number | null;
  edited_by_user: boolean;
  generated_at: string | null;
  created_at: string | null;
  updated_at: string | null;
}

export interface RelationshipCreate {
  from_table: string;
  from_column: string;
  to_table: string;
  to_column: string;
  kind: RelationshipKind;
  confidence?: number;
}

export interface RelationshipPatch {
  kind?: RelationshipKind;
}

export interface MetadataJob {
  id: string;
  user_id: string;
  status: MetadataJobStatus;
  progress: number;
  message: string | null;
  error: string | null;
  started_at: string | null;
  finished_at: string | null;
  aca_execution_name: string | null;
  created_at: string | null;
  updated_at: string | null;
}

export interface MetadataGenerationResponse {
  job: MetadataJob;
}

export interface ColumnValuesResponse {
  table: string;
  column: string;
  values: (string | number | boolean)[];
}

/* ─────────────────────────────────────────────
   FilterSpec types (mirrors backend models/filters.py)
───────────────────────────────────────────── */

export interface ColumnRef {
  table: string;
  column: string;
}

export type TimePreset =
  | 'last_7_days'
  | 'last_14_days'
  | 'last_30_days'
  | 'last_60_days'
  | 'last_90_days'
  | 'ytd'
  | 'mtd';

export interface TimeRange {
  from: string; // ISO date
  to: string;   // ISO date
}

export interface TimeFilter {
  column_ref: ColumnRef;
  preset?: TimePreset;
  range?: TimeRange;
}

export interface CategoricalFilter {
  kind: 'categorical';
  column_ref: ColumnRef;
  op: 'in' | 'not_in';
  values: (string | number | boolean)[];
}

export interface NumericFilter {
  kind: 'numeric';
  column_ref: ColumnRef;
  op: 'between';
  min?: number;
  max?: number;
}

export type Filter = CategoricalFilter | NumericFilter;

export interface FilterSpec {
  time?: TimeFilter;
  filters: Filter[];
}

File: backend/src/fastapi_app/models/metadata.py
"""
Pydantic models for the dashboard metadata layer.

These describe the LLM-generated, user-editable semantic schema that drives
native dashboard filtering. See ``docs/native_dashboard_filtering.md`` §3.

The codebase treats ``user_id`` as the tenant id (user = tenant).
"""

from __future__ import annotations

from datetime import datetime
from enum import Enum
from typing import Any

from pydantic import BaseModel, ConfigDict, Field


# ---------------------------------------------------------------------------
# Enums (mirror the Postgres enums in 6-metadata-tables.sql)
# ---------------------------------------------------------------------------


class SemanticType(str, Enum):
    categorical = "categorical"
    numeric = "numeric"
    temporal = "temporal"
    identifier = "identifier"
    measure = "measure"
    unknown = "unknown"


class RelationshipKind(str, Enum):
    many_to_one = "many_to_one"
    one_to_one = "one_to_one"
    many_to_many = "many_to_many"


class MetadataJobStatus(str, Enum):
    pending = "pending"
    running = "running"
    succeeded = "succeeded"
    failed = "failed"
    cancelled = "cancelled"


# ---------------------------------------------------------------------------
# Table metadata
# ---------------------------------------------------------------------------


class TableMetadata(BaseModel):
    """One row in ``tenant_table_metadata``."""

    model_config = ConfigDict(from_attributes=True)

    user_id: str
    table_name: str
    description: str | None = None
    primary_date_column: str | None = None
    grain: str | None = None
    generated_at: datetime | None = None
    edited_by_user: bool = False
    created_at: datetime | None = None
    updated_at: datetime | None = None


class TableMetadataPatch(BaseModel):
    """Partial update payload for a single table metadata row."""

    description: str | None = None
    primary_date_column: str | None = None
    grain: str | None = None


# ---------------------------------------------------------------------------
# Column metadata
# ---------------------------------------------------------------------------


class ColumnMetadata(BaseModel):
    """One row in ``tenant_column_metadata``."""

    model_config = ConfigDict(from_attributes=True)

    user_id: str
    table_name: str
    column_name: str
    data_type: str
    semantic_type: SemanticType = SemanticType.unknown
    description: str | None = None
    is_filterable: bool = True
    cardinality: int | None = None
    sample_values: list[Any] | None = None
    generated_at: datetime | None = None
    edited_by_user: bool = False
    created_at: datetime | None = None
    updated_at: datetime | None = None


class ColumnMetadataPatch(BaseModel):
    """Partial update payload for a single column metadata row."""

    semantic_type: SemanticType | None = None
    description: str | None = None
    is_filterable: bool | None = None


# ---------------------------------------------------------------------------
# Relationships
# ---------------------------------------------------------------------------


class Relationship(BaseModel):
    """One row in ``tenant_table_relationships``."""

    model_config = ConfigDict(from_attributes=True)

    user_id: str
    from_table: str
    from_column: str
    to_table: str
    to_column: str
    kind: RelationshipKind
    confidence: float | None = Field(default=None, ge=0.0, le=1.0)
    edited_by_user: bool = False
    generated_at: datetime | None = None
    created_at: datetime | None = None
    updated_at: datetime | None = None


class RelationshipCreate(BaseModel):
    """User-supplied edge to insert into ``tenant_table_relationships``."""

    from_table: str = Field(..., min_length=1)
    from_column: str = Field(..., min_length=1)
    to_table: str = Field(..., min_length=1)
    to_column: str = Field(..., min_length=1)
    kind: RelationshipKind
    confidence: float | None = Field(default=1.0, ge=0.0, le=1.0)


class RelationshipPatch(BaseModel):
    """Partial update for an existing edge (only ``kind`` is mutable today)."""

    kind: RelationshipKind | None = None


# ---------------------------------------------------------------------------
# Generation job
# ---------------------------------------------------------------------------


class MetadataJob(BaseModel):
    """One row in ``tenant_metadata_jobs``."""

    model_config = ConfigDict(from_attributes=True)

    id: str
    user_id: str
    status: MetadataJobStatus
    progress: float = 0.0
    message: str | None = None
    error: str | None = None
    started_at: datetime | None = None
    finished_at: datetime | None = None
    aca_execution_name: str | None = None
    created_at: datetime | None = None
    updated_at: datetime | None = None


class MetadataGenerationResponse(BaseModel):
    """Response returned by ``POST /api/metadata/generate``."""

    job: MetadataJob
